src/sentry/runner/commands/top.py

```diff
@@ -0,0 +1,440 @@
+
+
+def get_ip_address():
+    "Attempt to determine a private or public IP address"
+    import socket
+
+    ip = socket.gethostbyname(socket.gethostname())
+    if not ip.startswith('127.'):
+        return ip
+
+    import fcntl
+    import struct
+
+    for ifname in 'eth0', 'eth1', 'eth2', 'bond0', 'bond1', 'bond2':
+        try:
+            s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+            return socket.inet_ntoa(fcntl.ioctl(
+                s.fileno(),
+                0x8915,  # SIOCGIFADDR
+                struct.pack('256s', ifname[:15])
+            )[20:24])
+        except Exception:
+            continue
+
+    return None
+
+
+class TimeSeries(object):
+    "A container for time series data by key and value pairs"
+    __slots__ = ('_keys',)
+
+    def __init__(self):
+        self._keys = defaultdict(TimeSeriesKey)
+
+    def incr(self, key, value):
+        self._keys[key].incr(value)
+
+    def keys(self):
+        return self._keys.keys()
+
+    def top(self, key):
+        return self._keys[key].items()
+
+    def __iter__(self):
+        for k in self.keys():
+            yield k, self.top(k)
+
+
+class TimeSeriesKey(object):
+    "A specific time series key which contains many values"
+    __slots__ = ('_values',)
+
+    def __init__(self):
+        self._values = defaultdict(TimeSeriesValue)
+
+    def incr(self, value):
+        self._values[value].incr()
+
+    def keys(self):
+        return self._values
+
+    def items(self):
+        for k, v in self._values.items():
+            if v.is_dead():
+                del self._values[k]
+            yield k, (v.avg_short(), v.avg_mid(), v.avg_long())
+
+
+class TimeSeriesValue(object):
+    "A specific time series value which containes 3 points, short, mid, long"
+    __slots__ = ('_points',)
+
+    def __init__(self):
+        self._points = [
+            TimeSeriesPoint(1),
+            TimeSeriesPoint(5),
+            TimeSeriesPoint(10),
+        ]
+
+    def incr(self):
+        p = self._points
+        p[0].incr()
+        p[1].incr()
+        p[2].incr()
+
+    def is_dead(self):
+        # Bucket is considered dead, if any of the single points
+        # are dead since they are N'Sync.
+        return self._points[0].is_dead()
+
+    def avg_short(self):
+        return self._points[0].avg()
+
+    def avg_mid(self):
+        return self._points[1].avg()
+
+    def avg_long(self):
+        return self._points[2].avg()
+
+
+class TimeSeriesPoint(object):
+    "A single time series point"
+    __slots__ = ('_times', '_buckets', '_size')
+
+    def __init__(self, size=1):
+        self._size = size
+        # deques should be 1 larger than expected since
+        # we need a window. The latest bucket is always being
+        # filled up, and the others are used for data.
+        self._times = deque([0], size + 1)
+        self._buckets = deque([0], size + 1)
+
+    def incr(self):
+        ts = int(time())
```

Comment:
Isn't that effectively what I have here with the `deque`? It's fixed size memory.